firedrake/interpolation.py

```diff
@@ -95,20 +93,17 @@ def __init__(self, expr, v,
                 between a VOM and its input ordering. Defaults to ``True`` which uses SF broadcast
                 and reduce operations.
         """
-        # Check function space
-        expr = ufl.as_ufl(expr)
-        if isinstance(v, functionspaceimpl.WithGeometry):
-            expr_args = extract_arguments(expr)
+        if isinstance(V, functionspaceimpl.WithGeometry):
+            # Need to create a Firedrake Argument so that it has a .function_space() method
+            expr_args = extract_arguments(ufl.as_ufl(expr))
             is_adjoint = len(expr_args) and expr_args[0].number() == 0
-            v = Argument(v.dual(), 1 if is_adjoint else 0)
+            V = Argument(V.dual(), 1 if is_adjoint else 0)
```

Comment:
Put 0 if it is not there, otherwise keep increasing the count
```suggestion
            arg_numbers = [a.number() for a in expr_args]
            number = 0 if len(expr_args) == 0 or min(arg_numbers) > 0 else max(arg_numbers) + 1
            V = Argument(V.dual(), number)            
```

Reply:
Similar fixes will be needed in UFL